### ISSUE TITLE
Internalise ECDH TLS 1.2 helper functions

### ISSUE THREAD

**Issue by daverodgman** [Mbed-TLS/mbedtls#4041]
The ECDH API currently exposes helper functions for reading & writing TLS 1.2 key shares. This format changes in TLS 1.3, and given that these functions don't seem to be useful to users, they should be made not part of the public API

In Mbed TLS 4.0, we won't use these functions internally either, so we can remove them.

Prerequisite: https://github.com/Mbed-TLS/mbedtls/issues/8147

**Comment by hannestschofenig** [comment]
This is a good change! 

Once turned into an internal API I would like to see the following change: we need to add an extra argument to the function call to distinuguish between TLS 1.3 and pre-TLS 1.3 use of the ECDH parameters, as I explained in https://github.com/hannestschofenig/mbedtls/pull/116

**Comment by gilles-peskine-arm** [comment]
This is a controversial change. The ECDH TLS-helper functions correspond to the FFDH (`dhm.h`) interface, which doesn't have a “normal” DH interface. I'm not convinced we should remove the TLS-helper ECDH functions if the only way to use FFDH is through TLS-helper functions. My preference would be to get rid of the FFDH TLS-helper functions, but to do that we need to create a normal FFDH interface, and I'm not sure this fits into our 3.0 timeframe.

**Comment by mpg** [comment]
I tend to agree with Gilles. Also:

- the ECDH functions that are tied to TLS (<= 1.2) are also the only ECDH API that makes use of the ECDH context (as opposed to requiring the caller to manage `mpi`s and `ecp_point`s), so I'm not sure removing them without a replacement is a good idea. (Past interactions with users suggest the context makes the API easier to use, and some users seem to use the TLS API outside TLS - it's just a convenient serialization format.)
- having the current functions present is not really an obstacle to supporting TLS 1.3 in the future. By the way, it's not clear to me that future TLS 1.3 ECDH helpers would belong to `ecdh.h` rather than be internal to the SSL module. Also, I think we were considering basing TLS 1.3 directly on PSA Crypto APIs, in which case we really wouldn't care what's in `ecdh.h`.
- more generally, the `mbedtls_xxx` crypto API is on its way out, so I'm not particularly inclined to put effort in improving it (unless the improvement is really obvious, low effort big win).

So I'm inclined to reject this for 3.0.

**Comment by gilles-peskine-arm** [comment]
Mostly subsumed by https://github.com/Mbed-TLS/mbedtls/issues/8663 : the next major version will make all of `ecdh.h` internal. After that, and after https://github.com/Mbed-TLS/mbedtls/issues/8147 that removes these functions from the TLS code, we can remove the unused ECDH functions.

**Comment by mpg** [comment]
Note: after https://github.com/Mbed-TLS/mbedtls/pull/8753 (on top of the issues you mentioned) we can actually remove `ecdh.c` completely from our code base (not just make it internal).

**Comment by gilles-peskine-arm** [comment]
Moot because `ecdh.h` as a whole is becoming an internal API.
